presto-raptorx/src/main/java/com/facebook/presto/raptorx/metadata/DatabaseMetadataWriter.java

```diff
@@ -0,0 +1,728 @@
+ * </ul>
+ * </ul>
+ * Rollback protocol:
+ * <ul>
+ * <li>Starting a rollback:</li>
+ * <ul>
+ * <li>Mark {@code active_commit} row as aborted</li>
+ * </ul>
+ * <li>Writing to the master:</li>
+ * <ul>
+ * <li>Verify {@code active_commit} row has correct {@code commitId}</li>
+ * <li>Perform write operation</li>
+ * </ul>
+ * <li>Writing to a shard:</li>
+ * <ul>
+ * <li>Lock {@code aborted_commit} row and verify it is no larger than the current commit</li>
+ * <li>Update {@code aborted_commit} row</li>
+ * <li>Perform write operation</li>
+ * </ul>
+ * <li>Finishing a rollback:</li>
+ * <ul>
+ * <li>Verify {@code active_commit} row</li>
+ * <li>Delete {@code active_commit} row</li>
+ * </ul>
+ * </ul>
+ */
+public class DatabaseMetadataWriter
+        implements MetadataWriter
+{
+    private final SequenceManager sequenceManager;
+    private final Database.Type databaseType;
+    private final Jdbi masterDbi;
+    private final List<Jdbi> shardDbi;
+
+    @Inject
+    public DatabaseMetadataWriter(SequenceManager sequenceManager, Database database, TypeManager typeManager)
+    {
+        this.sequenceManager = requireNonNull(sequenceManager, "sequenceManager is null");
+        this.databaseType = requireNonNull(database.getType(), "database.type is null");
+
+        Jdbi masterDbi = createJdbi(database.getMasterConnection());
+        masterDbi.registerRowMapper(new ColumnInfo.Mapper(typeManager));
+        this.masterDbi = masterDbi;
+
+        this.shardDbi = database.getShards().stream()
+                .map(shard -> createJdbi(shard.getConnection()))
+                .collect(toImmutableList());
+    }
+
+    @Override
+    public void recover()
+    {
+        // start rollback
+        ActiveCommit activeCommit = masterDbi.inTransaction(handle -> {
+            MasterWriterDao dao = handle.attach(MasterWriterDao.class);
+
+            // acquire global lock
+            getLockedCurrentCommitId(dao);
+
+            // abort current commit
+            ActiveCommit commit = dao.getActiveCommit();
+            if (commit == null) {
+                return null;
+            }
+            if (!commit.isRollingBack()) {
+                dao.abortActiveCommit();
+            }
+
+            // rollback master
+            dao.rollback(commit.getCommitId());
+
+            return commit;
+        });
+
+        // skip if no rollback needed
+        if (activeCommit == null) {
+            return;
+        }
+
+        long commitId = activeCommit.getCommitId();
+        RollbackInfo rollbackInfo = activeCommit.getRollbackInfo();
+
+        // rollback shards
+        for (Jdbi shardDbi : shardDbi) {
+            shardDbi.useTransaction(handle -> {
+                ShardWriterDao dao = handle.attach(ShardWriterDao.class);
+
+                // acquire shard lock
+                long abortedCommitId = getLockedAbortedCommitId(dao);
+
+                // skip if no rollback needed
+                if (abortedCommitId > commitId) {
```

Comment:
I have 2 questions on this line:
1  given the current codes, I don't see how this can happen. We have database lock during commit and rollback. We cannot have:
t1 fails, commitID=1
t2 fails, commitID=2  // this cannot happen, t2 cannot even enters beginCommit()
t2 rollback 

2  If we are protecting against some corner case, "return" may not be what we intend to do, since its per shard. "continue" and "return " after the for loop?

Reply:
We might crash in the middle of a rollback and end up processing the same shard multiple times. This check is an optimization to skip the rollback for this shard if it has already been performed. It would be safe to remove this, but could end up setting the row to the existing and running the rest of the rollback code unnecessarily.

The `aborted_commit` table is the write lock for the shard. We might start a transaction that writes to a shard, have it fail, then perform the rollback while the write to the shard is still in flight (perhaps on a different machine after the coordinator appeared to crash). The write lock in the master database can't protect the shard since they are different databases.